# Colecciones

## Listas

Las listas son colecciones de datos que nos permitirá retener valores en un único elemento. Las listas poseen métodos para poder añadir, quitar, insertar o eliminar elementos. Una de las características más importante de las listas es su capacidad de extraer valores y sublistas a través de índices.

In [1]:
a = []

a

[]

In [2]:
a.append(1)
a.append(2)
a.append(3)
a.append(4)

a

[1, 2, 3, 4]

In [3]:
a[0] # 1

1

In [4]:
a[2] # 3

3

In [5]:
a.pop() # 4

4

In [6]:
a

[1, 2, 3]

In [7]:
a.pop() # 3

3

In [8]:
a

[1, 2]

In [9]:
frutas = ["manzana", "pera", "mango", "papaya", "melón"]

frutas

['manzana', 'pera', 'mango', 'papaya', 'melón']

In [10]:
last_fruta = frutas.pop()

last_fruta

'melón'

In [11]:
frutas

['manzana', 'pera', 'mango', 'papaya']

In [12]:
mango = frutas.pop(2)

mango

'mango'

In [13]:
frutas

['manzana', 'pera', 'papaya']

In [15]:
frutas.insert(2, "güayaba")

frutas

['manzana', 'pera', 'güayaba', 'papaya']

In [16]:
frutas.insert(1, "pera")

frutas

['manzana', 'pera', 'pera', 'güayaba', 'papaya']

In [17]:
frutas.remove("pera") # quita primer coincidencia

frutas

['manzana', 'pera', 'güayaba', 'papaya']

In [18]:
frutas.insert(3, "pera")

frutas

['manzana', 'pera', 'güayaba', 'pera', 'papaya']

In [19]:
frutas.remove("pera") # quita primer coincidencia

frutas

['manzana', 'güayaba', 'pera', 'papaya']

In [20]:
frutas.remove("melón")

ValueError: list.remove(x): x not in list

In [24]:
if "papaya" in frutas:
    frutas.remove("papaya")
    print(frutas)

['manzana', 'pera']


In [25]:
if frutas.count("güayaba") > 0:
    frutas.remove("güayaba")
    print(frutas)

In [28]:
import random

nums = []

for i in range(10):
    n = random.randint(1, 10)
    nums.append(n)
    
nums

[2, 2, 4, 6, 5, 7, 4, 6, 9, 5]

In [29]:
while nums.count(4) > 1:
    nums.remove(4)
    
nums

[2, 2, 6, 5, 7, 4, 6, 9, 5]

In [30]:
a = [1, 2, 3, 5, 7, 9]

# append - añade un elemento al final de la lista
a.append(10) # [1, 2, 3, 5, 7, 9, 10]

# pop - quita un elemento del final de la lista
a.pop() # [1, 2, 3, 5, 7, 9]

# pop - quita un elemento en el índice marcado * error si el índice no existe
a.pop(3) # [1, 2, 3, 7, 9]

# insert - inserta un elemento en el índice marcado * error si el índice es mayor al números de elementos o menor a cero
a.insert(3, 6) # [1, 2, 3, 6, 7, 9]

# remove - quita el primer elemento encontrado * error si no existe dentro de la lista
a.remove(7) # [1, 2, 3, 6, 9]

a

[1, 2, 3, 6, 9]

### Ejercicio - Generar una lista con los primeros 20 números de la secuencia de fibonacci

Los primeros dos números de fibonacci son 1 y 1, y el siguiente se calcula como la suma de los dos anteriores.

In [31]:
fibi = [1,1]

while len(fibi) < 20:
  n = len (fibi)
  s = fibi[n-1] + fibi [n-2]
  fibi.append (s)

print (fibi)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765]


## Índices y Sublistas

Los índices son las posiciones dentro de una secuencia (una lista o tupla), los cuáles comienzan marcar las posiciones desde `0` y podemos incluso maneajar índices negativos.

In [32]:
fibi = [1,1]

while len(fibi) < 20:
  s = fibi[-1] + fibi[-2]
  fibi.append (s)

print (fibi)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765]


In [35]:
a = range(1, 10) # [1, 2, 3, 4, 5, 6, 7, 8, 9]

a

range(1, 10)

In [39]:
a[0] # 1

1

In [38]:
a[6] # 7

7

In [40]:
a[-1] # 9

9

In [41]:
a[-2] # 8

8

In [42]:
a[-3] # 7

7

In [43]:
a[-9] # 1

1

In [44]:
a[-10] # error

IndexError: range object index out of range

In [47]:
# El último valor no es devuelto en índices
a[4:6] # [5, 7]

range(5, 7)

In [51]:
a = ["alberto", "bety", "carlos", "diana", "eduardo", "faby", "gus", "hugo", "isa"]

a

['alberto', 'bety', 'carlos', 'diana', 'eduardo', 'faby', 'gus', 'hugo', 'isa']

In [52]:
a[2] # carlos

'carlos'

In [53]:
a[2:5] # 2, 3, 4 ["carlos", "diana", "eduardo"]

['carlos', 'diana', 'eduardo']

In [54]:
a[3:7] # 3, 4, 5, 6 ["diana", "eduardo", "faby", "gus"]

['diana', 'eduardo', 'faby', 'gus']

In [55]:
a[3:] # ['diana', 'eduardo', 'faby', 'gus', 'hugo', 'isa']

['diana', 'eduardo', 'faby', 'gus', 'hugo', 'isa']

In [56]:
a[:5] # ['alberto', 'bety', 'carlos', 'diana', 'eduardo']

['alberto', 'bety', 'carlos', 'diana', 'eduardo']

## Ejercicio - Partición de listas

Partir una lista de N elementos en sublistas de 3 elementos, partiendo del elemento inicial y hasta el último.

In [58]:
import random

x = []

for i in range(20):
    x.append(random.random())
    
x

[0.6097159820376516,
 0.6141304733873578,
 0.28029978405255485,
 0.40232742123546095,
 0.8968812374226506,
 0.3631281228451958,
 0.13245154581209495,
 0.7661729899458928,
 0.6430217512021982,
 0.38899755971489225,
 0.4563111266023785,
 0.49742691197468114,
 0.7919896577899962,
 0.07397495805968701,
 0.7601849816733824,
 0.6500345210631555,
 0.5210400191455647,
 0.6278419534765117,
 0.0162361686488649,
 0.9438789957918131]

In [59]:
n = len(x)

n

20

In [64]:
from math import ceil, floor

ceil(n / 3) # 20 / 3 = 6.66666... ceil - redondeo hacia arriba (7), floor - redondeo hacia abajo (6)

7

In [65]:
floor(20 / 3)

6

In [70]:
# 0, 3, 6, 9, ..., 18

# floor(n / 3) nos indica cuántas sublistas completas de 3 caben en nuestra lista (respuesta hay 6 ternas completas)
# 3 * floor(n / 3) = 3 * 6 = 18, es decir, pordemo avanzar de 3 en 3 hasta el índice 18 sin tocarlo, es decir, 0, 3, 6, 9, 12, 15

for i in range(0, 3 * floor(n / 3), 3): # range(a, b, s) a, a + s, a + 2s, a + 3s, ..., b - 1, range(0, 18, 3) = 0, 3, 6, 9, 12, 15
    print(i)

0
3
6
9
12
15


In [71]:
# 0, 3, 6, 9, ...

for i in range(0, 3 * floor(n / 3), 3):
    print(x[i], x[i + 1], x[i + 2])

0.6097159820376516 0.6141304733873578 0.28029978405255485
0.40232742123546095 0.8968812374226506 0.3631281228451958
0.13245154581209495 0.7661729899458928 0.6430217512021982
0.38899755971489225 0.4563111266023785 0.49742691197468114
0.7919896577899962 0.07397495805968701 0.7601849816733824
0.6500345210631555 0.5210400191455647 0.6278419534765117


### Ejemplo - Codificación de colores en imágenes

Generalmente las imágenes son guardadas en archivos codificados dónde cada 3 valores representan un pixel RGB, esto quiere decir que el archivo en realidad es una lista de números, dónde cada 3 números hay color codificado. Vamos a recuperar de una lista de valores entre 0 y 255, los colores RGB que representa cada pixel. El valor 0 se refiere a una ausencia de color y 255 el color más intenso, entonces, si recuperamos (0, 255, 0) significaría ausencia de rojo (Red) y verde (Green) y el color azul (Blue) más intenso.

Vamos a generar una imagen de 8x8 pixeles, entonces tendríamos un total de 8x8x3 valores o 192 valores en una lista, cuyo rango es de 0 a 255. Entonces hacemo lo siguiente.

In [75]:
import random

values = []

for i in range(8*8*3):
    x = random.randint(0, 256) # un valor aleatorio entre 0 y 255
    values.append(x)
    
values[:9]

[72, 227, 224, 174, 25, 182, 131, 220, 33]

Podemos observar que los primeros 3 valores son 72, 227 y 244, los cuáles son el RGB de un pixel en la imagen de 8x8, es decir, el primer pixel.

In [79]:
for i in range(0, 3 * floor(len(values) / 3), 3):
    print(i, values[i:i+3]) # recuperamos la sublista desde la posición `i` (el índice de la terna o pixel) hasta el siguiente índice (3 más)

0 [72, 227, 224]
3 [174, 25, 182]
6 [131, 220, 33]
9 [243, 248, 92]
12 [209, 80, 126]
15 [67, 212, 156]
18 [110, 231, 208]
21 [203, 236, 179]
24 [218, 208, 212]
27 [172, 252, 85]
30 [220, 138, 182]
33 [14, 93, 215]
36 [125, 181, 15]
39 [231, 117, 11]
42 [84, 87, 127]
45 [147, 4, 166]
48 [99, 236, 134]
51 [14, 186, 128]
54 [238, 211, 75]
57 [31, 31, 171]
60 [209, 16, 102]
63 [44, 58, 119]
66 [205, 108, 226]
69 [65, 254, 241]
72 [227, 249, 30]
75 [165, 256, 96]
78 [189, 60, 146]
81 [55, 204, 160]
84 [144, 210, 143]
87 [49, 212, 29]
90 [0, 128, 22]
93 [240, 168, 91]
96 [74, 74, 199]
99 [95, 82, 7]
102 [109, 5, 141]
105 [16, 31, 10]
108 [228, 154, 54]
111 [240, 74, 218]
114 [224, 201, 214]
117 [33, 83, 210]
120 [182, 75, 74]
123 [215, 131, 71]
126 [2, 64, 87]
129 [166, 111, 94]
132 [231, 137, 97]
135 [178, 15, 146]
138 [244, 249, 27]
141 [166, 183, 204]
144 [215, 154, 204]
147 [186, 151, 151]
150 [29, 242, 203]
153 [214, 40, 0]
156 [148, 189, 225]
159 [230, 57, 73]
162 [173, 140, 174]
165 

## Tuplas

Las tuplas son paquetes de valores similares a las listas pero de tamaño fijo, y no podemos reemplazar sus valores. Las tuplas están diseñadas para manejar varios valores de múltiples tipos de manera ordenada a través de un mismo elemento, el sentido es por lo tanto, empaquetar valores para transportarlos y manipularlos fácilmente.

In [80]:
t = (1, True, "pedro")

t

(1, True, 'pedro')

In [81]:
t[0]

1

In [82]:
t[1]

True

In [83]:
t[2]

'pedro'

In [84]:
t[0] = 2 # error

TypeError: 'tuple' object does not support item assignment

Las tuplas pueden ser desacopladas (desmpaquetadas) en el número de variables igual número de elementos en la tupla (dimensión).

In [85]:
a, b, c = t # a, b, c = (1, True, "pedro"), a = 1, b = True, c = "pedro"

In [86]:
a

1

In [87]:
b

True

In [88]:
c

'pedro'

In [89]:
a, b = (1, 100)

print(a, b)

1 100


Una tupla con `n` valores se considera una n-tupla, es decir, una tupla de 2 valores acoplados se considera una 2-tupla

In [90]:
p = (1, 5)

p

(1, 5)

In [91]:
x, y = p

f"x: {x}, y: {y}"

'x: 1, y: 5'

La combinación de tuplas y listas nos permite crear listas complejas las cuáles pueden retener varios valores (una tupla por cada elemento) y manipular los elementos de forma compleja. Por ejemplo, si quisieramos retener los valores de un polígono, podríamos guardar en una lista la coordena de cada punto, como una 2-tupla (x, y).

In [93]:
poligono = [
    (0, 0),
    (1, 0),
    (1, 1),
    (0, 1)
]

poligono

[(0, 0), (1, 0), (1, 1), (0, 1)]

De esta manera almacenamos en la lista cada elemento acoplado. para acceder a los elementos podemos hacerlo por índice o iterandólos.

In [94]:
poligono[1] # (1, 0)

(1, 0)

In [95]:
x1, y1 = poligono[1]

print(x1, y1)

1 0


In [96]:
for p in poligono:
    print(p)

(0, 0)
(1, 0)
(1, 1)
(0, 1)


In [99]:
for p in poligono:
    x, y = p # desacoplamos la 2-tupla en 2 variables
    print(f"{x}, {y}")

0, 0
1, 0
1, 1
0, 1


Al recorrer una lista de tuplas, podemos desacoplar los elementos directamente ahorrando una línea de código.

In [98]:
for x, y in poligono: # desacopla cada 2-tupla en x, y al recorrer cada 2-tupla del polígono
    print(f"{x}, {y}")

0, 0
1, 0
1, 1
0, 1


### Ejemplo: Calcular el centro geométrico de un polígono

Un polígono es una lista de coordenadas (vértices) y el centro geométrico es el punto promedio de las x's y las y's respectivamente.

In [100]:
poligono = [
    (0, 0),
    (1, 0),
    (1, 1),
    (0, 1)
]

xs = 0
ys = 0

for x, y in poligono:
    xs += x
    ys += y
    
xp = xs / len(poligono)
yp = ys / len(poligono)

f"xp: {xp}, yp: {yp}"

'xp: 0.5, yp: 0.5'

Podemos a partir de dos listas de la misma longitud, crear una lista acoplada (o una lista de tuplas) usando la función zip.

In [102]:
X = [1, 2, 3, 4, 5, 6]
Y = [3, 6, 4, 7, 8, 2]

# Con esta operación acoplamos columnas (series)
p = list( zip(X, Y) ) # acoplamos X y Y, para tener una lista [(x, y)]

p

[(1, 3), (2, 6), (3, 4), (4, 7), (5, 8), (6, 2)]

También podemos desacoplar una lista acoplada (lista de tuplas) en n-listas, según las n-tuplas. Usandp zip* creamos la operación inversa al acoplamiento.

In [103]:
data = [
    ("pedro", "hombre"), # 2-tupla
    ("ana", "mujer"),
    ("luis", "hombre"),
    ("bety", "mujer"),
]

data

[('pedro', 'hombre'), ('ana', 'mujer'), ('luis', 'hombre'), ('bety', 'mujer')]

In [108]:
# Con esta operación desacoplamos columnas
nombres, generos = zip(*data) # 2-tuplas (2-columnas, 2-series)

# Si queremos recuperar los ejes como listas entonces los debemos convertir
#nombres, generos = list(map(list, zip(*data))) # 2-listas (2-columnas, 2-series)

print(nombres)
print(generos)

['pedro', 'ana', 'luis', 'bety']
['hombre', 'mujer', 'hombre', 'mujer']


### Ejemplo - Generar los valores de la serie del coseno

In [105]:
from math import cos, pi

X = [-pi, -pi/2, 0, pi/2, pi]

# Esta sintáxis es para una lista generada a partir de una iteración
# [valor for valor in secuencia]
# genera una lista con el `valor` para cada `valor` en la secuencia
Y = [cos(x) for x in X] # genera una lista con el valor `cos(x)` para cada `x` en la lista de X

print(X)
print(Y)

[-3.141592653589793, -1.5707963267948966, 0, 1.5707963267948966, 3.141592653589793]
[-1.0, 6.123233995736766e-17, 1.0, 6.123233995736766e-17, -1.0]


In [107]:
P = list(zip(X, Y))

P

[(-3.141592653589793, -1.0),
 (-1.5707963267948966, 6.123233995736766e-17),
 (0, 1.0),
 (1.5707963267948966, 6.123233995736766e-17),
 (3.141592653589793, -1.0)]

Generalmente el acoplado de listas zip, se utiliza para recorrer varias series al mismo tiempo.

In [109]:
X = [1, 4, 5, 6, 9]
Y = [3, 6, 5, 8, 10]

for x, y in zip(X, Y):
    print(f"{x}, {y}, {x * y}")

1, 3, 3
4, 6, 24
5, 5, 25
6, 8, 48
9, 10, 90


In [110]:
for i in range(len(X)):
    x = X[i]
    y = Y[i]
    print(f"{x}, {y}, {x * y}")

1, 3, 3
4, 6, 24
5, 5, 25
6, 8, 48
9, 10, 90


In [111]:
for i, x in enumerate(X):
    print(f"{i}: {x}")

0: 1
1: 4
2: 5
3: 6
4: 9


## Diccionarios

Los diccionarios son valores acoplados igual que las tuplas, pero en dónde el índice (la clave) es definida por el usuario, una tupla acopla n-valores secuencialmente, por ejemplo, `("pedro", "homre")`, dónde en índice `0` se almacena `"pedro"` y en el índe `1` se almacena `"hombre"`. Sin embargo, una forma más natural de acoplar datos es colocarlos en un diccionario, cuyas claves conozcamos perfectamente y no tengamos problemas al recuperar o cambiar la información asociada a la clave.

In [112]:
# usando tuplas acoplamos la información de una persona como: persona = ("pedro", "hombre")
# recuperar un valor requiere conocer su índice, lo cuál puede no ser natural (¿Por qué índice 1 es hombre?)

# usando diccionarios acoplamos la información de una persona, a través de claves
persona = { "nombre": "pedro", "genero": "hombre" }
# recuperar un valor requiere conocer su clave, esto lo vuelve más natural (hombre está almacenado en la clave genero)

persona

{'nombre': 'pedro', 'genero': 'hombre'}

In [116]:
persona = {
    "nombre": "ana",
    "edad": 21,
    "genero": "mujer",
    "casado": True,
    "hijos": 4
}

persona

{'nombre': 'ana', 'edad': 21, 'genero': 'mujer', 'casado': True, 'hijos': 4}

In [114]:
persona["nombre"] # ana

'ana'

In [117]:
persona["casado"] # True

True

Si las claves no están dentro del diccionario, debemo preguntar antes de operarlas.

In [118]:
persona["nacimiento"] # error

KeyError: 'nacimiento'

In [119]:
if "nacimiento" in persona:
    print(persona["nacimiento"])
else:
    print("La clave 'nacimiento' no está en el diccionario")

La clave 'nacimiento' no está en el diccionario


In [120]:
persona["nacimiento"] = "1999/04/17"

if "nacimiento" in persona:
    print(persona["nacimiento"])
else:
    print("La clave 'nacimiento' no está en el diccionario")

1999/04/17


También mediante la clave podemos modificar el valor asociado a esa clave.

In [121]:
persona["edad"] = 20

persona

{'nombre': 'ana',
 'edad': 20,
 'genero': 'mujer',
 'casado': True,
 'hijos': 4,
 'nacimiento': '1999/04/17'}

Una aplicación muy común en los diccionarios es usarlos mediante listas de diccionarios, al igual que las listas de tuplas, tener listas de diccionarios nos van permitir iterar información compleja.

In [122]:
personas = [
    {
        "nombre": "pedro",
        "genero": "hombre"
    },
    {
        "nombre": "ana",
        "genero": "mujer"
    },
    {
        "nombre": "juan",
        "genero": "hombre"
    },
    {
        "nombre": "tito",
        "genero": "hombre"
    },
    {
        "nombre": "monica",
        "genero": "mujer"
    },
]

personas

[{'nombre': 'pedro', 'genero': 'hombre'},
 {'nombre': 'ana', 'genero': 'mujer'},
 {'nombre': 'juan', 'genero': 'hombre'},
 {'nombre': 'tito', 'genero': 'hombre'},
 {'nombre': 'monica', 'genero': 'mujer'}]

In [123]:
for persona in personas:
    print(persona["nombre"])

pedro
ana
juan
tito
monica


Cuándo trabajamos con listas, podemos generar otras listar a partir de la lista original, utilizando las listas generadas `[value for value in sequence]`.

In [125]:
nombres = [persona["nombre"] for persona in personas] # genera una lista con persona["nombre"] para cada persona en la lista de personas

nombres

['pedro', 'ana', 'juan', 'tito', 'monica']

In [127]:
generos = [persona["genero"] for persona in personas]

generos

['hombre', 'mujer', 'hombre', 'hombre', 'mujer']

In [128]:
for nombre, genero in zip(nombres, generos):
    print(f"{nombre} es {genero}")

pedro es hombre
ana es mujer
juan es hombre
tito es hombre
monica es mujer


In [129]:
for persona in personas:
    nombre = persona["nombre"]
    genero = persona["genero"]
    print(f"{nombre} es {genero}")

pedro es hombre
ana es mujer
juan es hombre
tito es hombre
monica es mujer


In [131]:
personas

[{'nombre': 'pedro', 'genero': 'hombre'},
 {'nombre': 'ana', 'genero': 'mujer'},
 {'nombre': 'juan', 'genero': 'hombre'},
 {'nombre': 'tito', 'genero': 'hombre'},
 {'nombre': 'monica', 'genero': 'mujer'}]

In [132]:
for nombre, genero in zip([p["nombre"] for p in personas], [p["genero"] for p in personas]):
    print(f"{nombre} ({genero})")

pedro (hombre)
ana (mujer)
juan (hombre)
tito (hombre)
monica (mujer)
